### DISCUSSION TITLE
how to merge/reuse cached images from different versions/deployments?

### DISCUSSION THREAD

**Discussion by hrvoj3e** [cantaloupe-project/cantaloupe#508] [Q&A] [answered]
Is it possible to merge caches (derivate, source, info) from multiple cantaloupe deployments?

Are caching hashes backward compatible? E.g. version 3.3.1 and 4.4.1?
How would I approach migrating multiple instances (different versions) to one instance?

**Comment by adolski** [comment, marked as answer]
> Is it possible to merge caches (derivate, source, info) from multiple cantaloupe deployments?

Yes, as long as the identifiers don't clash between the two deployments. For example, if deployment 1 knows of an `image.jpg` that is 500x500 pixels, and deployment 2 knows of a different `image.jpg` that is 750x750 pixels, that will be a problem.

> Are caching hashes backward compatible? E.g. version 3.3.1 and 4.4.1?

The cache key format varies between the different cache implementations, but basically there are two formats:

1. Identifier MD5 checksum used for infos. This is generally stable across versions.
2. Identifier MD5 checksum + serialized operation list MD5 checksum used for derivative images. This may change if the operation list implementation (`e.i.l.c.operation.OperationList`) changes. (More specifically, if you take FilesystemCache as an example, a derivative image filename is in the format `{identifier checksum}_{op list checksum}.{format}`. In this case, the 2nd checksum will change.)

Additionally, the info serialization has changed a few times. These changes should be backward compatible so that current software versions can understand previous serializations, but the inverse may or may not be the case (it doesn't really get tested).

There isn't a firm policy on maintaining stable cache keys. I try not to break them unnecessarily, but if there is some kind of redesign that I perceive to be in the best interests of the software, then I will usually go ahead with it and break them. This may happen across new minor versions (x.x), but not bugfix versions (x.x.x).

Stale cache files shouldn't cause any problems by remaining around, other than wasting space, and should eventually get cleaned up by the cache worker or by a PurgeInvalidFromCache command.
